### DISCUSSION TITLE
Symfony6 stateful and stateless access to a shared resource

### DISCUSSION THREAD

**Discussion by olia-bn** [symfony/symfony#48216] [Q&A] [answered]
What is the proper way of enabling both stateful and stateless access to a resource in Symfony6?

In my app, API can be acessed an authentication header, but in Symfony5, an user authenticated in a stateful manner was also granted access, which was handy as I used the API for a lot of my frontend components (like tables and Vue components) seamlesly.

I am unable to figure out if I can configure my firewalls in Symfony6 in the same way. They can't share a context since one of them is stateless. 

To solve the issue, I made a mixin that is always aware of a valid authentication token, but I'd much rather have either authentication method be available and enable access to both stateful and stateless requests.

**Comment by MatTheCat** [comment, marked as answer]
A firewall is not limited to a single authenticator. From https://symfony.com/doc/current/security.html#the-firewall:

> the firewall defines which parts of your application are secured and _how_ your users will be able to authenticate

So if you want to secure resources the same way, you need a single firewall. If there are two authentication mechanisms to access the resources, this firewall will configure two authenticators.

**Comment by olia-bn** [comment]
Okay so if I understand you correctly, what I need is configuring multiple authenticators via the `custom_authenticators` config option for my firewall.

**Reply by MatTheCat** [reply]
In fact I’m not even sure you need two authenticators in your case: if the firewall is stateful Symfony should check the session automatically :thinking:
